allennlp/data/vocabulary.py

```diff
@@ -285,6 +287,48 @@ def from_dataset(cls,
                           max_vocab_size=max_vocab_size,
                           non_padded_namespaces=non_padded_namespaces)
 
+    @classmethod
+    def from_params(cls, params: Params, dataset=None):
+        """
+        There are two possible ways to build a vocabulary; from a
+        pre-existing dataset, using :func:`Vocabulary.from_dataset`, or
+        from a pre-saved vocabulary, using :func:`Vocabulary.from_files`.
+        This method wraps both of these options, allowing their specification
+        from a ``Params`` object, generated from a JSON configuration file.
+
+        Parameters
+        ----------
+        params: Params, required.
+        dataset: Dataset, optional.
+            If ``params`` doesn't contain a ``vocabulary_directory`` key,
+            the ``Vocabulary`` can be built directly from a ``Dataset``.
+
+        Returns
+        -------
+        A ``Vocabulary``.
+        """
+        vocabulary_directory = params.pop("vocabulary_directory", None)
+        if not vocabulary_directory and not dataset:
+            raise ConfigurationError("You must provide either a Params object containing a "
+                                     "vocab_directory key or a Dataset to build a vocabulary from.")
+        if vocabulary_directory and dataset:
+            logger.warning("from_params was passed both a vocabulary_directory and a Dataset. This "
```

Comment:
This will happen every time you pass this flag during training, because the training code passes the dataset.  I'd instead have the warning say something like "Constructing vocabulary from files instead of from the dataset", and doing it as `logger.info` should be fine.